File: src/ainu_mcp/research.py
"""Composed lookup: gather everything we know about a word in one call.

This is the killer tool — instead of Claude making 5 separate calls to draft a
glossary entry, it asks for `entry_research(word)` once and gets:

- All script renditions (Latin / Katakana / Cyrillic)
- Existing glossary entries (any category)
- Dictionary hits across all dictionaries
- Up to N corpus example sentences
"""

from __future__ import annotations

from typing import Any

from . import corpus, dictionaries, glossary, script


def entry_research(
    word: str,
    *,
    corpus_limit: int = 8,
    dict_limit: int = 12,
    glossary_limit: int = 10,
) -> dict[str, Any]:
    scripts = script.all_scripts(word)
    try:
        syllables = script.separate_syllables(scripts.get("latn", word))
    except Exception as e:
        syllables = [f"(error: {e})"]

    # build a set of query forms — Latin and Kana both useful for corpus/dict hits
    forms = {word, scripts.get("latn", ""), scripts.get("kana", "")} - {""}

    glossary_hits: list[dict[str, Any]] = []
    seen_rows: set[tuple[str, int]] = set()
    for form in forms:
        try:
            for hit in glossary.search(form, limit=glossary_limit):
                key = (hit["category"], hit["row"])
                if key in seen_rows:
                    continue
                seen_rows.add(key)
                glossary_hits.append(hit)
        except Exception as e:
            glossary_hits.append({"error": f"glossary search failed for '{form}': {e}"})

    dict_hits: list[dict[str, Any]] = []
    for form in forms:
        try:
            dict_hits.extend(dictionaries.lookup(form, limit=dict_limit))
        except Exception as e:
            dict_hits.append({"error": f"dictionary lookup failed for '{form}': {e}"})

    corpus_hits: list[dict[str, Any]] = []
    for form in forms:
        try:
            corpus_hits.extend(corpus.search(form, lang="ain", limit=corpus_limit))
        except Exception as e:
            corpus_hits.append({"error": f"corpus search failed for '{form}': {e}"})

    return {
        "query": word,
        "scripts": scripts,
        "syllables": syllables,
        "glossary": glossary_hits[:glossary_limit],
        "dictionaries": dict_hits[:dict_limit],
        "corpus": corpus_hits[:corpus_limit],
    }

File: worker/src/tools/research.ts
/** Composed entry_research (port of ainu_mcp/research.py). */
import { z } from "zod";
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import type { Env } from "../types.js";
import { corpusSearch } from "../corpus-source.js";
import { glossarySearch } from "./glossary.js";
import { lookupEntries } from "./dictionaries.js";
import { allScripts, separateSyllables } from "./script.js";
import { jsonResult } from "./helpers.js";

export function registerResearchTools(server: McpServer, env: Env): void {
  server.tool(
    "entry_research",
    "One-shot lookup composing: script conversions + syllable separation + existing glossary entries + multi-dictionary lookups + corpus examples. Call this when drafting a new glossary entry or vetting an existing one — one call returns everything needed to make a high-quality edit.",
    {
      word: z.string(),
      corpus_limit: z.number().int().default(8),
      dict_limit: z.number().int().default(12),
      glossary_limit: z.number().int().default(10),
    },
    async ({ word, corpus_limit, dict_limit, glossary_limit }) => {
      const scripts = allScripts(word);
      let syllables: string[];
      try {
        syllables = separateSyllables(scripts.latn ?? word);
      } catch (e) {
        syllables = [`(error: ${e instanceof Error ? e.message : String(e)})`];
      }

      const forms = [...new Set([word, scripts.latn ?? "", scripts.kana ?? ""])].filter(Boolean);

      const glossaryHits: unknown[] = [];
      const seen = new Set<string>();
      for (const form of forms) {
        try {
          for (const hit of await glossarySearch(env, { query: form, limit: glossary_limit })) {
            const key = `${hit.category} ${hit.row}`;
            if (seen.has(key)) continue;
            seen.add(key);
            glossaryHits.push(hit);
          }
        } catch (e) {
          glossaryHits.push({ error: `glossary search failed for '${form}': ${e instanceof Error ? e.message : String(e)}` });
        }
      }

      const dictHits: unknown[] = [];
      for (const form of forms) {
        try {
          dictHits.push(...(await lookupEntries(env, { word: form, limit: dict_limit })));
        } catch (e) {
          dictHits.push({ error: `dictionary lookup failed for '${form}': ${e instanceof Error ? e.message : String(e)}` });
        }
      }

      const corpusHits: unknown[] = [];
      for (const form of forms) {
        try {
          const rows = await corpusSearch(env, { query: form, lang: "ain", limit: corpus_limit });
          corpusHits.push(
            ...rows.map((r) => ({
              id: r.id,
              text: r.text,
              translation: r.translation,
              dialect: r.dialect,
              author: r.author,
              collection: r.collection,
              document: r.document,
              uri: r.uri,
            })),
          );
        } catch (e) {
          corpusHits.push({ error: `corpus search failed for '${form}': ${e instanceof Error ? e.message : String(e)}` });
        }
      }

      return jsonResult({
        query: word,
        scripts,
        syllables,
        glossary: glossaryHits.slice(0, glossary_limit),
        dictionaries: dictHits.slice(0, dict_limit),
        corpus: corpusHits.slice(0, corpus_limit),
      });
    },
  );
}
